README.md

```diff
@@ -1,48 +1,147 @@
+//     Name:         "consul",
+//     TargetSystem: "aws",
+//   },
+//   Subdir: "modules/consul-cluster",
+// },
+```
+
+## Other Module Address Formats
+
+Modules can also be sourced from [other sources](https://www.terraform.io/language/modules/sources)
+and these other sources (outside of Terraform Registry)
+have different address formats, such as `./local` or
+`github.com/hashicorp/example`.
+
+This library does _not_ recognize such other address formats
+and it will return error upon parsing these.
+
+## Ambiguous Provider Addresses
+
+Qualified addresses with namespace (such as `hashicorp/aws`)
+are used exclusively in all recent versions (`0.14+`) of Terraform.
+If you only work with Terraform `v0.14.0+` configuration/output, you may
+safely ignore the rest of this section and related part of the API.
+
+There are a few types of ambiguous addresses you may comes accross:
+
+ - Terraform `v0.12` uses "namespace-less address", such as `aws`.
+ - Terraform `v0.13` may use `-` as a placeholder for the unknown namespace,
+   resulting in address such as `-/aws`.
+ - Terraform `v0.14+` _configuration_ still allows ambiguous providers
+   through `provider "<NAME>" {}` block _without_ corresponding
+   entry inside `required_providers`, but these providers are always
+   resolved as `hashicorp/<NAME>` and all JSON outputs only use that
+   resolved address.
+
+Both ambiguous address formats are accepted by `ParseProviderSource()`
+
+```go
+pAddr, err := ParseProviderSource("aws")
+if err != nil {
+	// deal with error
+}
+
+// pAddr == Provider{
+//   Type:      "aws",
+//   Namespace: UnknownProviderNamespace,    // "?"
+//   Hostname:  DefaultProviderRegistryHost, // "registry.terraform.io"
+// }
+pAddr.HasKnownNamespace() // == false
+pAddr.IsLegacy() // == false
+```
+```go
+pAddr, err := ParseProviderSource("-/aws")
+if err != nil {
+	// deal with error
+}
 
-A legacy address is by itself (without more context) ambiguous.
-For example `aws` may represent either the official `hashicorp/aws`
-or just any custom-built provider called `aws`.
+// pAddr == Provider{
+//   Type:      "aws",
+//   Namespace: LegacyProviderNamespace,     // "-"
+//   Hostname:  DefaultProviderRegistryHost, // "registry.terraform.io"
+// }
+pAddr.HasKnownNamespace() // == true
+pAddr.IsLegacy() // == true
+```
 
-Such ambiguous address can be produced by Terraform `<=0.12`. You can
-just use `ImpliedProviderForUnqualifiedType` if you know for sure
-the address was produced by an affected version.
+However `NewProvider()` will panic if you pass an empty namespace
+or any placeholder indicating unknown namespace.
 
-If you do not have that context you should parse the string via
-`ParseRawProviderSourceString` and then check `addr.IsLegacy()`.
+```go
+NewProvider(DefaultProviderRegistryHost, "aws", "")  // panic
+NewProvider(DefaultProviderRegistryHost, "aws", "-") // panic
+NewProvider(DefaultProviderRegistryHost, "aws", "?") // panic
```

Comment:
Are the namespace and provider name parameters flipped here?

Reply:
Yes, good catch - thanks!